Fix Subs._eval_derivative function for higher order derivative

gxyd opened the pull request:
Fixes #12005 , one of the test cases(only one) taken from diofant.
If this fix has any similarity with the code of skirpichev that is probably a coincidence. I have only used the test case from his PR.
<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below.>

gxyd committed:
Fix Subs._eval_derivative function for higher order derivative

Fixes #12005, one of the test cases(only one) taken from diofant

jksuom commented on sympy/core/function.py:
```diff
@@ -1738,10 +1738,9 @@ def _eval_subs(self, old, new):
-                    self.variables, self.point).doit() for arg,
-                    point in zip(self.variables, self.point) ])
+        return Add(*[point.diff(s)*self.func(self.expr.diff(var), self.variables, self.point).doit() + \
+                self.func(self.expr.diff(s) if s != var else S.Zero, self.variables, self.point).doit() \
```
It looks like this term may occur several times.
```
In [42]: e = Subs(Derivative(f(x), x), (y, z), (y, z))

In [43]: print(e.diff(x))
2*Derivative(f(x), x, x)

In [45]: e = Subs(Derivative(f(x), x), (y, z, t), (y, z, t))

In [46]: print(e.diff(x))
3*Derivative(f(x), x, x)
```
The backslashes are probably not needed.

jksuom commented:
The derivation variable `s` can appear in `e = Subs(expr, vars, points)` in two places that may contribute to the derivative, in `expr` and in `points`. If it is also among the `vars`, then it could be replaced, in principle, by a dummy in both `expr` and `vars` before derivation. In this case, only the appearances of `s` in `points` affect the derivative.

Hence, if `s` is one of `vars`, the derivative of `expr` with respect to `s` should not be included.

gxyd committed:
Replace derivative of expr with 0, when s is variable
if s is one of vars, the derivative of expr with respect to s
should not be included.

gxyd commented:
Okay, I made those changes in case where `s` may be present in one of `variables`. For the first point, I have ousted the second term from `Add` (I forgot to add its test case, perhaps will add in next commit).

gxyd committed:
Use Add class as common for two terms, add one more tests

Idea for common class from diofant/diofant 3f7ca353 commit

gxyd commented:
@jksuom please review.

jksuom commented:
This looks good to me.
